systemtest/infra_pubsublite.go

```diff
@@ -0,0 +1,94 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package systemtest
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"path/filepath"
+
+	"github.com/hashicorp/terraform-exec/tfexec"
+)
+
+var (
+	googleProject string
+	googleRegion  string
+)
+
+// PubSubLiteConfig for PubSubLite topics, subscriptions, reservations
+type PubSubLiteConfig struct {
+	// TFPath is the path where the terraform files are present.
+	TFPath string
+	// Project is the GCP Project name.
+	Project string
+	// Region is the GCP region.
+	Region string
+	// Topics (and subscriptions) to create.
+	Topics []string
+	// ReservationSuffix to use.
+	ReservationSuffix string
+}
+
+func newPubSubLiteConfig(topics ...string) PubSubLiteConfig {
+	return PubSubLiteConfig{
+		Topics:  topics,
+		Project: googleProject,
+		Region:  googleRegion,
+		TFPath:  filepath.Join("tf", "pubsublite"),
+	}
+}
+
+// ProvisionPubSubLite provisions the PubSubLite resources. The specified
+// terraform path must accept at least these variables:
+// - project
+// - region
+// - suffix (used to suffix the reservation)
+// - topics
+func ProvisionPubSubLite(ctx context.Context, cfg PubSubLiteConfig) error {
+	if cfg.ReservationSuffix == "" {
+		cfg.ReservationSuffix = persistentSuffix
+	}
+	logger().Infof("provisioning PubSubLite infrastructure with config: %+v", cfg)
+	tf, err := NewTerraform(ctx, cfg.TFPath)
+	if err != nil {
+		return err
+	}
+	if err := tf.Init(ctx, tfexec.Upgrade(true)); err != nil {
+		return err
+	}
+
+	jsonTopics, err := json.Marshal(cfg.Topics)
+	if err != nil {
+		return err
+	}
+	projectVar := tfexec.Var(fmt.Sprintf("project=%s", cfg.Project))
+	regionVar := tfexec.Var(fmt.Sprintf("region=%s", cfg.Region))
+	suffixVar := tfexec.Var(fmt.Sprintf("suffix=%s", cfg.ReservationSuffix))
+	topicsVar := tfexec.Var(fmt.Sprintf("topics=%s", jsonTopics))
+	// Ensure terraform destroy runs once per TF path.
+	RegisterDestroy(cfg.TFPath, func() {
+		logger().Info("destroying provisioned PubSubLite infrastructure...")
+		tf.Destroy(ctx, projectVar, regionVar, suffixVar, topicsVar)
+	})
+	if err := tf.Apply(ctx, projectVar, regionVar, suffixVar, topicsVar); err != nil {
```

Comment:
**issue**: we should run `tf.Validate` before applying to make sure there are no issues with the file

Reply:
If there are, terraform will fail anyway. So I'm not sure `tf.Validate()` gives us anything extra here.

Reply:
It would fail but there's no validation contrary to `terraform plan` or `terraform validate` so you might bump into strange/unclear error messages.
I had to deal with a similar issue in the apm-server smoke tests and ended up adding `terraform validate` so make sure the configuration was valid.
I don't feel strongly about this so I'll leave it up to you. We can always revisit this later if we ever have the need to.

Reply:
I think we can leave it for now. If it becomes a problem we can always add `tf.Validate()`